# DAG
 


## 1.Instalar POLARS

%pip install polars pandas pyarrow

## 2.Importar librerias 

In [2]:
import polars as pl

## TAREA 1: Cargar Datos 

In [3]:
ventas_feriado= pl.read_csv("./Datos/holidays_events.csv")
precio_petroleo= pl.read_csv("./Datos/oil.csv")
ventas_54= pl.read_csv("./Datos/stores.csv")
ventas_principal= pl.read_csv("./Datos/train.csv")
transacciones= pl.read_csv("./Datos/transactions.csv")


## TAREA 2: EDA INICIAL

* Generar un diagnóstico inicial por cada archivo  
* Generar un reporte 

In [24]:
datasets = {
    "ventas_principal": ventas_principal,
    "ventas_54": ventas_54,
    "ventas_feriado": ventas_feriado,
    "precio_petroleo": precio_petroleo,
    "transacciones": transacciones
}

import json

reporte = {}

for nombre, d in datasets.items():

    filas, columnas = d.shape

    porcentaje_nulos = (
        d.null_count()
        .transpose(include_header=True, header_name="columna", column_names=["nulos"])
        .with_columns(
            (pl.col("nulos") / filas * 100).round(2).alias("porcentaje")
        )
        .filter(pl.col("nulos") > 0)
        .sort("porcentaje", descending=True)
    )

    total = d.height
    unicos = d.unique().height
    duplicados = total - unicos

    reporte[nombre] = {
        "filas": filas,
        "columnas": columnas,
        "tipos_datos": {str(col): str(tipo) for col, tipo in d.schema.items()},
        "nulos": porcentaje_nulos.to_dicts() if porcentaje_nulos.height > 0 else "Sin valores nulos",
        "duplicados": duplicados,
        "porcentaje_duplicados": round((duplicados / total) * 100, 2)
    }


# Guardar reporte JSON
with open("reporte_datasets.json", "w", encoding="utf-8") as archivo:
    json.dump(
        reporte,
        archivo,
        indent=4,
        ensure_ascii=False
    )

print("Reporte JSON generado correctamente")


Reporte JSON generado correctamente


## TAREA 3: Limpieza de datos

### Estandarización de fechas 

In [5]:
columna_fecha = 'date'

for nombre, d in datasets.items(): 
    try:
        datasets[nombre] = d.with_columns(
            pl.col(columna_fecha).str.to_date(format="%Y-%m-%d", strict=False)
        )
        
        rango = datasets[nombre].select([
            pl.col(columna_fecha).min().alias("minimo"),
            pl.col(columna_fecha).max().alias("maximo")
        ])
        
        print(f"Dataset: {nombre}")    
        print(f"  Rango de fechas: {rango['minimo'][0]} hasta {rango['maximo'][0]}")
        print("-" * 45)

    except pl.ColumnNotFoundError:
        print(f"Dataset: {nombre}") 
        print("No existe la columna fecha")
        print("-" * 45)


Dataset: ventas_principal
  Rango de fechas: 2013-01-01 hasta 2017-08-15
---------------------------------------------
Dataset: ventas_54
No existe la columna fecha
---------------------------------------------
Dataset: ventas_feriado
  Rango de fechas: 2012-03-02 hasta 2017-12-26
---------------------------------------------
Dataset: precio_petroleo
  Rango de fechas: 2013-01-01 hasta 2017-08-31
---------------------------------------------
Dataset: transacciones
  Rango de fechas: 2013-01-01 hasta 2017-08-15
---------------------------------------------


C:\Users\ASUS\AppData\Local\Temp\ipykernel_11808\133085479.py:18: DeprecationWarning: accessing `ColumnNotFoundError` from the top-level `polars` module was deprecated in version 1.0.0. Import it directly from the `polars.exceptions` module instead, e.g.: `from polars.exceptions import ColumnNotFoundError`
  except pl.ColumnNotFoundError:


### Eliminación de Duplicados

In [ ]:
for nombre, d in datasets.items():
    filas_antes = d.height
    datasets[nombre] = d.unique(maintain_order=True)
    filas_despues = datasets[nombre].height
    print(f"{nombre}: {filas_antes - filas_despues} duplicados eliminados")


ventas_principal: 0 duplicados eliminados
ventas_54: 0 duplicados eliminados
ventas_feriado: 0 duplicados eliminados
precio_petroleo: 0 duplicados eliminados
transacciones: 0 duplicados eliminados


In [7]:
oil = datasets["precio_petroleo"]
rango_completo = pl.DataFrame({
    "date": pl.date_range(oil["date"].min(), oil["date"].max(), interval="1d", eager=True)
})
oil_completo = (
    rango_completo.join(oil, on="date", how="left")
    .sort("date")
    .with_columns(pl.col("dcoilwtico").interpolate())
    .with_columns(pl.col("dcoilwtico").fill_null(strategy="backward").fill_null(strategy="forward"))
)
print(f"precio_petroleo:\nFilas \nAntes: {oil.height}\nDespues: {oil_completo.height}")
print(f"\nNulos \nAntes: {oil['dcoilwtico'].null_count()}\nDespues: {oil_completo['dcoilwtico'].null_count()}")
datasets["precio_petroleo"] = oil_completo

print()

for nombre, d in datasets.items():
    if nombre == "precio_petroleo":
        continue
    filas_antes, columnas_antes = d.shape
    for col, dtype in d.schema.items():
        if d[col].null_count() == 0:
            continue
        valor = d[col].median() if dtype.is_numeric() else d[col].mode()[0]
        d = d.with_columns(pl.col(col).fill_null(valor))
    filas_despues, columnas_despues = d.shape
    print(f"{nombre}:\nFilas \nAntes: {filas_antes}\nDespues: {filas_despues}")
    print(f"Columnas \nAntes: {columnas_antes}\nDespues: {columnas_despues}\n")
    datasets[nombre] = d

precio_petroleo:
Filas 
Antes: 1218
Despues: 1704

Nulos 
Antes: 43
Despues: 0

ventas_principal:
Filas 
Antes: 3000888
Despues: 3000888
Columnas 
Antes: 6
Despues: 6

ventas_54:
Filas 
Antes: 54
Despues: 54
Columnas 
Antes: 5
Despues: 5

ventas_feriado:
Filas 
Antes: 350
Despues: 350
Columnas 
Antes: 6
Despues: 6

transacciones:
Filas 
Antes: 83488
Despues: 83488
Columnas 
Antes: 3
Despues: 3



### Correccion de tipo de datos

In [ ]:

datasets["ventas_principal"] = datasets["ventas_principal"].with_columns([
    pl.col("family").cast(pl.Categorical),
])

datasets["ventas_54"] = datasets["ventas_54"].with_columns([
    pl.col("city").cast(pl.Categorical),
    pl.col("state").cast(pl.Categorical),
    pl.col("type").cast(pl.Categorical),
])

datasets["ventas_feriado"] = datasets["ventas_feriado"].with_columns([
    pl.col("type").cast(pl.Categorical),
    pl.col("locale").cast(pl.Categorical),
    pl.col("locale_name").cast(pl.Categorical),
])

for nombre, d in datasets.items():
    print(f"{nombre}: {d.schema}")

ventas_principal: Schema({'id': Int64, 'date': Date, 'store_nbr': Int64, 'family': Categorical, 'sales': Float64, 'onpromotion': Int64})
ventas_54: Schema({'store_nbr': Int64, 'city': Categorical, 'state': Categorical, 'type': Categorical, 'cluster': Int64})
ventas_feriado: Schema({'date': Date, 'type': Categorical, 'locale': Categorical, 'locale_name': Categorical, 'description': String, 'transferred': Boolean})
precio_petroleo: Schema({'date': Date, 'dcoilwtico': Float64})
transacciones: Schema({'date': Date, 'store_nbr': Int64, 'transactions': Int64})


## Tarea 4: Union de datasets

In [9]:
ventas_feriado_unica = datasets["ventas_feriado"].group_by("date").first()

ventas_completo = (
    datasets["ventas_principal"]
    .join(datasets["ventas_54"], on="store_nbr", how="left")
    .join(datasets["transacciones"], on=["date", "store_nbr"], how="left")
    .join(datasets["precio_petroleo"], on="date", how="left")
    .join(ventas_feriado_unica, on="date", how="left")
    .with_columns(pl.col("transactions").fill_null(0))
)
ventas_completo.head()

id,date,store_nbr,family,sales,onpromotion,city,state,type,cluster,transactions,dcoilwtico,type_right,locale,locale_name,description,transferred
i64,date,i64,cat,f64,i64,cat,cat,cat,i64,i64,f64,cat,cat,cat,str,bool
0,2013-01-01,1,"""AUTOMOTIVE""",0.0,0,"""Quito""","""Pichincha""","""D""",13,0,93.14,"""Holiday""","""National""","""Ecuador""","""Primer dia del ano""",false
1,2013-01-01,1,"""BABY CARE""",0.0,0,"""Quito""","""Pichincha""","""D""",13,0,93.14,"""Holiday""","""National""","""Ecuador""","""Primer dia del ano""",false
2,2013-01-01,1,"""BEAUTY""",0.0,0,"""Quito""","""Pichincha""","""D""",13,0,93.14,"""Holiday""","""National""","""Ecuador""","""Primer dia del ano""",false
3,2013-01-01,1,"""BEVERAGES""",0.0,0,"""Quito""","""Pichincha""","""D""",13,0,93.14,"""Holiday""","""National""","""Ecuador""","""Primer dia del ano""",false
4,2013-01-01,1,"""BOOKS""",0.0,0,"""Quito""","""Pichincha""","""D""",13,0,93.14,"""Holiday""","""National""","""Ecuador""","""Primer dia del ano""",false


### Evaluación de Dataset Unificado


In [ ]:
print(f"Dataset Ventas_completo")

filas, columnas = ventas_completo.shape

print(f"Filas: {filas:,}")
print(f"Columnas: {columnas}")

print("Tipos de dato:")
print(ventas_completo.schema)

# Valores nulos
nulos = (
    ventas_completo.null_count()
    .transpose(include_header=True, header_name="columna", column_names=["nulos"])
    .with_columns(
        (pl.col("nulos") / filas * 100).round(2).alias("porcentaje")
    )
    .filter(pl.col("nulos") > 0)
    .sort("porcentaje", descending=True)
)

print("Nulos por columna:")
print(nulos if nulos.height > 0 else "Sin valores nulos")

# Duplicados
duplicados = filas - ventas_completo.unique().height
print(f"Filas duplicadas: {duplicados} ({duplicados / filas * 100:.2f}%)")

# Rango de fechas
columna_fecha = "date"

if columna_fecha in ventas_completo.columns:

    if ventas_completo.schema[columna_fecha] == pl.Date:
        fechas = ventas_completo.select(pl.col(columna_fecha))
    else:
        fechas = ventas_completo.select(
            pl.col(columna_fecha).str.to_date(strict=False)
        )

    print(f"Rango de fechas: {fechas.min()[0,0]} a {fechas.max()[0,0]}")

else:
    print("No tiene columna de fecha")

print("-" * 40)

Dataset: shape: (3_000_888, 17)
┌─────────┬────────────┬───────────┬────────────┬───┬──────────┬───────────┬───────────┬───────────┐
│ id      ┆ date       ┆ store_nbr ┆ family     ┆ … ┆ locale   ┆ locale_na ┆ descripti ┆ transferr │
│ ---     ┆ ---        ┆ ---       ┆ ---        ┆   ┆ ---      ┆ me        ┆ on        ┆ ed        │
│ i64     ┆ date       ┆ i64       ┆ cat        ┆   ┆ cat      ┆ ---       ┆ ---       ┆ ---       │
│         ┆            ┆           ┆            ┆   ┆          ┆ cat       ┆ str       ┆ bool      │
╞═════════╪════════════╪═══════════╪════════════╪═══╪══════════╪═══════════╪═══════════╪═══════════╡
│ 0       ┆ 2013-01-01 ┆ 1         ┆ AUTOMOTIVE ┆ … ┆ National ┆ Ecuador   ┆ Primer    ┆ false     │
│         ┆            ┆           ┆            ┆   ┆          ┆           ┆ dia del   ┆           │
│         ┆            ┆           ┆            ┆   ┆          ┆           ┆ ano       ┆           │
│ 1       ┆ 2013-01-01 ┆ 1         ┆ BABY CARE  ┆ … ┆ Natio

### LIMPIEZA DESPUÉS DE UNIFICAR 

In [ ]:
#  Manejo de nulos 
ventas_completo = ventas_completo.with_columns([
    pl.col("transactions").fill_null(0),
    pl.col("dcoilwtico")
        .interpolate()
        .fill_null(strategy="forward")
        .fill_null(strategy="backward"),
    pl.col("onpromotion").fill_null(0),
    pl.col("sales").fill_null(pl.col("sales").median()),

    pl.col("type_right").fill_null("Normal"),
    pl.col("locale").fill_null("National"),
    pl.col("locale_name").fill_null("Ecuador"),
    pl.col("description").fill_null("Sin feriado"),
    pl.col("transferred").fill_null(False)
])


# Eliminación de duplicados

if duplicados > 0:
    ventas_completo = ventas_completo.unique(maintain_order=True)

# VERIFICACIÓN
print("\nEstado final:")
print(f"Filas: {ventas_completo.height:,}")
print(f"Columnas: {ventas_completo.width}")
nulos_finales = (
    ventas_completo.null_count()
    .transpose(include_header=True, header_name="columna", column_names=["nulos"])
    .filter(pl.col("nulos") > 0)
)
print("\nNulos después de limpieza:")
print(nulos_finales if nulos_finales.height > 0 else "Sin valores nulos")
print("=" * 50)


Estado final:
Filas: 3,000,888
Columnas: 17

Nulos después de limpieza:
Sin valores nulos


## TAREA 6: EDA profundo (sobre datos limpios y consolidados)

 Ventas generales

- Distribución de ventas por familia de producto: ¿qué categorías concentran el mayor volumen?

In [11]:
# display(ventas_completo)
ventas_por_familia = (ventas_completo.group_by("family")
    .agg(pl.col("sales").sum().alias("ventas_totales"))
    .sort("ventas_totales", descending=True)
    .with_columns(
        (pl.col("ventas_totales") / pl.col("ventas_totales").sum() * 100).round(2).alias("porcentaje")
    )
)
display(ventas_por_familia)

family,ventas_totales,porcentaje
cat,f64,f64
"""GROCERY I""",3.4346e8,31.99
"""BEVERAGES""",2.16954486e8,20.21
"""PRODUCE""",1.2270e8,11.43
"""CLEANING""",9.7521289e7,9.08
"""DAIRY""",6.4487709e7,6.01
…,…,…
"""MAGAZINES""",266359.0,0.02
"""HARDWARE""",103470.0,0.01
"""HOME APPLIANCES""",41601.0,0.0


> **Interpretación:**
> 
> Podemos ver que la categoría que más se concentra es "GROCERY I", llegando a un 
> total de 31.99% de ventas, seguido de "BEVERAGES" con 20.21% de ventas y en tercer 
> lugar tenemos a "PRODUCE" con 11.43% de ventas.

- Ventas totales por tienda y ranking de las 10 tiendas con mayor y menor venta.

In [12]:
ventas_por_tienda = (ventas_completo.group_by("store_nbr")
    .agg(pl.col("sales").sum().alias("ventas_totales"))
    .sort("ventas_totales", descending=True)
)

print("Top 10 tiendas con mayor venta:")
print(ventas_por_tienda.head(10))

print("\nTop 10 tiendas con menor venta:")
print(ventas_por_tienda.tail(10).sort("ventas_totales"))


Top 10 tiendas con mayor venta:
shape: (10, 2)
┌───────────┬────────────────┐
│ store_nbr ┆ ventas_totales │
│ ---       ┆ ---            │
│ i64       ┆ f64            │
╞═══════════╪════════════════╡
│ 44        ┆ 6.2088e7       │
│ 45        ┆ 5.4498e7       │
│ 47        ┆ 5.0948e7       │
│ 3         ┆ 5.0482e7       │
│ 49        ┆ 4.3420e7       │
│ 46        ┆ 4.1896e7       │
│ 48        ┆ 3.5933e7       │
│ 51        ┆ 3.2911e7       │
│ 8         ┆ 3.0494e7       │
│ 50        ┆ 2.8653e7       │
└───────────┴────────────────┘

Top 10 tiendas con menor venta:
shape: (10, 2)
┌───────────┬────────────────┐
│ store_nbr ┆ ventas_totales │
│ ---       ┆ ---            │
│ i64       ┆ f64            │
╞═══════════╪════════════════╡
│ 52        ┆ 2.6962e6       │
│ 22        ┆ 4.0902e6       │
│ 32        ┆ 5.9518e6       │
│ 30        ┆ 7.3821e6       │
│ 35        ┆ 7.6767e6       │
│ 26        ┆ 7.7551e6       │
│ 42        ┆ 8.9458e6       │
│ 21        ┆ 9.2555e6       │
│ 10  

> **Interpretación:**
> 
> Podemos ver que la tienda con mayor ventas que ha tenido es la 44 y la que menor ventas ha tenido es la 52

- Ventas promedio por ciudad y provincia.

In [13]:
ventas_por_ciudad_provincia = (
    ventas_completo.group_by(["state", "city"])
    .agg(pl.col("sales").mean().round(2).alias("venta_promedio"))
    .sort("venta_promedio", descending=True)
)
    
print("Ventas promedio por ciudad, agrupado por provincia:")
print(ventas_por_ciudad_provincia)

Ventas promedio por ciudad, agrupado por provincia:
shape: (22, 3)
┌────────────┬───────────┬────────────────┐
│ state      ┆ city      ┆ venta_promedio │
│ ---        ┆ ---       ┆ ---            │
│ cat        ┆ cat       ┆ f64            │
╞════════════╪═══════════╪════════════════╡
│ Pichincha  ┆ Quito     ┆ 556.58         │
│ Pichincha  ┆ Cayambe   ┆ 509.71         │
│ Tungurahua ┆ Ambato    ┆ 362.63         │
│ Guayas     ┆ Daule     ┆ 345.28         │
│ Loja       ┆ Loja      ┆ 339.38         │
│ …          ┆ …         ┆ …              │
│ Manabi     ┆ El Carmen ┆ 198.98         │
│ Cotopaxi   ┆ Latacunga ┆ 190.58         │
│ Guayas     ┆ Playas    ┆ 138.14         │
│ Manabi     ┆ Manta     ┆ 125.17         │
│ Pastaza    ┆ Puyo      ┆ 73.6           │
└────────────┴───────────┴────────────────┘


> **Interpretación:**
> 
> Podemos ver que la ciudad de Quito en la provincia de Pichincha tiene una venta promedio mayor con un total de
> $556.58 

- Evolución temporal de ventas: tendencia mensual y anual entre 2013 y 2017.

In [14]:
ventas_mensual = (
    ventas_completo
    .with_columns([
        pl.col("date").dt.year().alias("anio"),
        pl.col("date").dt.month().alias("mes"),
    ])
    .group_by(["anio", "mes"])
    .agg(pl.col("sales").sum().alias("ventas_totales"))
    .sort(["anio", "mes"])
)
print("Evolución mensual:")
print(ventas_mensual)

ventas_anual = (
    ventas_completo
    .with_columns(pl.col("date").dt.year().alias("anio"))
    .group_by("anio")
    .agg(pl.col("sales").sum().alias("ventas_totales"))
    .sort("anio")
)
print("\nEvolución anual:")
print(ventas_anual)

Evolución mensual:
shape: (56, 3)
┌──────┬─────┬────────────────┐
│ anio ┆ mes ┆ ventas_totales │
│ ---  ┆ --- ┆ ---            │
│ i32  ┆ i8  ┆ f64            │
╞══════╪═════╪════════════════╡
│ 2013 ┆ 1   ┆ 1.0328e7       │
│ 2013 ┆ 2   ┆ 9.6590e6       │
│ 2013 ┆ 3   ┆ 1.1428e7       │
│ 2013 ┆ 4   ┆ 1.0993e7       │
│ 2013 ┆ 5   ┆ 1.1598e7       │
│ …    ┆ …   ┆ …              │
│ 2017 ┆ 4   ┆ 2.5895e7       │
│ 2017 ┆ 5   ┆ 2.6912e7       │
│ 2017 ┆ 6   ┆ 2.5683e7       │
│ 2017 ┆ 7   ┆ 2.7011e7       │
│ 2017 ┆ 8   ┆ 1.2433e7       │
└──────┴─────┴────────────────┘

Evolución anual:
shape: (5, 2)
┌──────┬────────────────┐
│ anio ┆ ventas_totales │
│ ---  ┆ ---            │
│ i32  ┆ f64            │
╞══════╪════════════════╡
│ 2013 ┆ 1.4042e8       │
│ 2014 ┆ 2.0947e8       │
│ 2015 ┆ 2.4088e8       │
│ 2016 ┆ 2.8865e8       │
│ 2017 ┆ 1.9422e8       │
└──────┴────────────────┘


> **Interpretación:**
> 
> Podemos ver que las ventas crecen año a año, desde "1.4042e8" en 2013 hasta un pico
> de 2.8865e8 en 2016. El valor más bajo de 2017 "1.9422e8" se debe a que el dataset 
> solo llega hasta agosto de ese año, no a una caída real en las ventas.


## Estacionalidad y feriados

- Impacto de feriados nacionales en el volumen de ventas: comparación días feriados vs días normales.

> **Interpretación:**
> 
> Podemos ver que el impacto del volumen de ventas tiene más impacto en los días normales.
> 
> 


- Ventas en los tres días previos y posteriores a feriados nacionales por familia de producto.

In [ ]:
    fechas_feriados_nacionales = (
        datasets["ventas_feriado"]
        .filter((pl.col("type") == "Holiday") & (pl.col("locale") == "National"))
        .select("date")
        .unique()
        .to_series()
        .to_list()
    )

    periodos_tiempo = []
    for fecha in fechas_feriados_nacionales:
        for offset in range(-3, 4):
            periodos_tiempo.append({
                "date": fecha.__class__.fromordinal(fecha.toordinal() + offset),
                "dias_relativo": offset,
                "feriado_ref": fecha
            })

    df_periodos = pl.DataFrame(periodos_tiempo)

    ventas_periodos_feriado = (
        ventas_completo
        .join(df_periodos, on="date", how="inner")
        .group_by(["family", "dias_relativo"])
        .agg(pl.col("sales").mean().round(2).alias("venta_promedio"))
        .sort(["family", "dias_relativo"])
    )
    print(ventas_periodos_feriado)

shape: (231, 3)
┌────────────┬───────────────┬────────────────┐
│ family     ┆ dias_relativo ┆ venta_promedio │
│ ---        ┆ ---           ┆ ---            │
│ cat        ┆ i64           ┆ f64            │
╞════════════╪═══════════════╪════════════════╡
│ AUTOMOTIVE ┆ -3            ┆ 6.71           │
│ AUTOMOTIVE ┆ -2            ┆ 6.83           │
│ AUTOMOTIVE ┆ -1            ┆ 6.64           │
│ AUTOMOTIVE ┆ 0             ┆ 6.71           │
│ AUTOMOTIVE ┆ 1             ┆ 7.22           │
│ …          ┆ …             ┆ …              │
│ SEAFOOD    ┆ -1            ┆ 20.56          │
│ SEAFOOD    ┆ 0             ┆ 21.99          │
│ SEAFOOD    ┆ 1             ┆ 23.65          │
│ SEAFOOD    ┆ 2             ┆ 23.03          │
│ SEAFOOD    ┆ 3             ┆ 22.44          │
└────────────┴───────────────┴────────────────┘


> **Interpretación:**
> 
> Podemos ver que el efecto del feriado varía por familia: AUTOMOTIVE se mantiene 
> estable, mientras que SEAFOOD aumenta notablemente en el feriado y los días 
> posteriores (de 20.56 a 23.65), mostrando mayor sensibilidad a los feriados.


- ¿Qué familias de productos son más sensibles a los feriados?

In [16]:
venta_normal_por_familia = (
    ventas_feriado_flag
    .filter(pl.col("tipo_dia") == "Normal")
    .group_by("family")
    .agg(pl.col("sales").mean().alias("venta_normal"))
)

venta_feriado_por_familia = (
    ventas_feriado_flag
    .filter(pl.col("tipo_dia") == "Feriado")
    .group_by("family")
    .agg(pl.col("sales").mean().alias("venta_feriado"))
)

sensibilidad_feriados = (
    venta_normal_por_familia
    .join(venta_feriado_por_familia, on="family", how="inner")
    .with_columns(
        ((pl.col("venta_feriado") - pl.col("venta_normal")) / pl.col("venta_normal") * 100)
        .round(2).alias("variacion_pct")
    )
    .sort("variacion_pct", descending=True)
)
print(sensibilidad_feriados)

NameError: name 'ventas_feriado_flag' is not defined

> **Interpretación:**
> 
> Podemos ver que "LIQUOR,WINE,BEER" y "HOME APPLIANCES" son las familias más 
> sensibles positivamente a los feriados, con aumentos de más del 10%. En cambio,
> familias como "FROZEN FOODS" y "CELEBRATION" muestran caídas de más del 20% en 
> feriados, lo que indica que no todas las familias se benefician de estas fechas.
> 


## Promociones

- Comparación de ventas con y sin promoción por familia de producto.

In [ ]:
ventas_promocion = (
    ventas_completo
    .with_columns(
        pl.when(pl.col("onpromotion") > 0)
        .then(pl.lit("Con promoción"))
        .otherwise(pl.lit("Sin promoción"))
        .alias("estado_promo")
    )
    .group_by(["family", "estado_promo"])
    .agg(pl.col("sales").mean().round(2).alias("venta_promedio"))
    .sort(["family", "estado_promo"])
)
print(ventas_promocion)

shape: (65, 3)
┌────────────────────────────┬───────────────┬────────────────┐
│ family                     ┆ estado_promo  ┆ venta_promedio │
│ ---                        ┆ ---           ┆ ---            │
│ cat                        ┆ str           ┆ f64            │
╞════════════════════════════╪═══════════════╪════════════════╡
│ AUTOMOTIVE                 ┆ Con promoción ┆ 13.24          │
│ AUTOMOTIVE                 ┆ Sin promoción ┆ 5.85           │
│ BABY CARE                  ┆ Con promoción ┆ 1.66           │
│ BABY CARE                  ┆ Sin promoción ┆ 0.11           │
│ BEAUTY                     ┆ Con promoción ┆ 8.29           │
│ …                          ┆ …             ┆ …              │
│ PRODUCE                    ┆ Sin promoción ┆ 792.1          │
│ SCHOOL AND OFFICE SUPPLIES ┆ Con promoción ┆ 44.53          │
│ SCHOOL AND OFFICE SUPPLIES ┆ Sin promoción ┆ 1.02           │
│ SEAFOOD                    ┆ Con promoción ┆ 35.43          │
│ SEAFOOD                

> **Interpretación:**
> 
> Podemos ver que las familias venden más cuando están en promoción.
> Por ejemplo, AUTOMOTIVE pasa de 5.85 a 13.24 y SCHOOL AND OFFICE SUPPLIES de 
> 1.02 a 44.53.


- ¿Las promociones incrementan las ventas? ¿En qué familias tienen mayor efecto?

In [ ]:
venta_sin_promo_por_familia = (
    ventas_promocion
    .filter(pl.col("estado_promo") == "Sin promoción")
    .select(["family", "venta_promedio"])
    .rename({"venta_promedio": "venta_sin_promo"})
)

venta_con_promo_por_familia = (
    ventas_promocion
    .filter(pl.col("estado_promo") == "Con promoción")
    .select(["family", "venta_promedio"])
    .rename({"venta_promedio": "venta_con_promo"})
)

efecto_promocion = (
    venta_sin_promo_por_familia
    .join(venta_con_promo_por_familia, on="family", how="inner")
    .with_columns(
        ((pl.col("venta_con_promo") - pl.col("venta_sin_promo")) / pl.col("venta_sin_promo") * 100)
        .round(2).alias("variacion_pct")
    )
    .sort("variacion_pct", descending=True)
)

print("¿Las promociones incrementan las ventas?")
print(efecto_promocion)

incremento_promedio = efecto_promocion["variacion_pct"].mean()
print(f"\nIncremento promedio general por promoción: {incremento_promedio:.2f}%")

print("\nTop 5 familias con mayor efecto de promoción:")
print(efecto_promocion.head(5))

¿Las promociones incrementan las ventas?
shape: (32, 4)
┌────────────────────────────┬─────────────────┬─────────────────┬───────────────┐
│ family                     ┆ venta_sin_promo ┆ venta_con_promo ┆ variacion_pct │
│ ---                        ┆ ---             ┆ ---             ┆ ---           │
│ cat                        ┆ f64             ┆ f64             ┆ f64           │
╞════════════════════════════╪═════════════════╪═════════════════╪═══════════════╡
│ SCHOOL AND OFFICE SUPPLIES ┆ 1.02            ┆ 44.53           ┆ 4265.69       │
│ BABY CARE                  ┆ 0.11            ┆ 1.66            ┆ 1409.09       │
│ PET SUPPLIES               ┆ 3.51            ┆ 15.84           ┆ 351.28        │
│ HOME AND KITCHEN II        ┆ 10.28           ┆ 41.16           ┆ 300.39        │
│ PRODUCE                    ┆ 792.1           ┆ 2438.22         ┆ 207.82        │
│ …                          ┆ …               ┆ …               ┆ …             │
│ CLEANING                   ┆ 

> **Interpretación:**
>
> Sí, las promociones aumentan bastante las ventas, en promedio un 289%.
> El efecto es mucho mayor en productos que la gente compra poco seguido, como útiles escolares o artículos para 
> bebés. En productos de consumo diario como huevos o verduras, la promoción ayuda, pero mucho menos.

## Petróleo y economía

- Correlación entre precio del petróleo y ventas totales mensuales.

In [ ]:
ventas_petroleo_mensual = (
    ventas_completo
    .with_columns([
        pl.col("date").dt.year().alias("anio"),
        pl.col("date").dt.month().alias("mes"),
    ])
    .group_by(["anio", "mes"])
    .agg([
        pl.col("sales").sum().alias("ventas_totales"),
        pl.col("dcoilwtico").mean().alias("precio_petroleo_promedio")
    ])
    .sort(["anio", "mes"])
)

correlacion = ventas_petroleo_mensual.select(
    pl.corr("ventas_totales", "precio_petroleo_promedio").alias("correlacion")
)
print(ventas_petroleo_mensual)
print(f"\nCorrelación petróleo vs ventas: {correlacion[0,0]:.4f}")

shape: (56, 4)
┌──────┬─────┬────────────────┬──────────────────────────┐
│ anio ┆ mes ┆ ventas_totales ┆ precio_petroleo_promedio │
│ ---  ┆ --- ┆ ---            ┆ ---                      │
│ i32  ┆ i8  ┆ f64            ┆ f64                      │
╞══════╪═════╪════════════════╪══════════════════════════╡
│ 2013 ┆ 1   ┆ 1.0328e7       ┆ 94.705484                │
│ 2013 ┆ 2   ┆ 9.6590e6       ┆ 95.431429                │
│ 2013 ┆ 3   ┆ 1.1428e7       ┆ 93.237419                │
│ 2013 ┆ 4   ┆ 1.0993e7       ┆ 91.804667                │
│ 2013 ┆ 5   ┆ 1.1598e7       ┆ 94.695968                │
│ …    ┆ …   ┆ …              ┆ …                        │
│ 2017 ┆ 4   ┆ 2.5895e7       ┆ 51.054833                │
│ 2017 ┆ 5   ┆ 2.6912e7       ┆ 48.574355                │
│ 2017 ┆ 6   ┆ 2.5683e7       ┆ 45.199333                │
│ 2017 ┆ 7   ┆ 2.7011e7       ┆ 46.495484                │
│ 2017 ┆ 8   ┆ 1.2433e7       ┆ 48.884667                │
└──────┴─────┴────────────────┴──────────

> **Interpretación:**
>
> Podemos ver una correlación negativa fuerte (-0.75): cuando el petróleo baja de 
> precio, las ventas suben. Esto se refleja entre 2013 y 2017, donde el petróleo 
> cae de 94.7 a 48.9 mientras las ventas casi se triplican.


- Identificación del lag temporal entre caída del petróleo y caída en ventas (período 2015-2016).

In [ ]:
periodo_analisis = (
    ventas_petroleo_mensual
    .filter((pl.col("anio") >= 2015) & (pl.col("anio") <= 2016))
    .sort(["anio", "mes"])
    .with_columns([
        pl.col("precio_petroleo_promedio").pct_change().alias("var_petroleo"),
        pl.col("ventas_totales").pct_change().alias("var_ventas"),
    ])
)
print(periodo_analisis)

for lag in range(1, 4):
    lag_df = periodo_analisis.with_columns(
        pl.col("var_petroleo").shift(lag).alias(f"var_petroleo_lag{lag}")
    )
    corr_lag = lag_df.select(
        pl.corr(f"var_petroleo_lag{lag}", "var_ventas").alias("correlacion")
    )
    print(f"Correlación con lag de {lag} mes(es): {corr_lag[0,0]:.4f}")

shape: (24, 6)
┌──────┬─────┬────────────────┬──────────────────────────┬──────────────┬────────────┐
│ anio ┆ mes ┆ ventas_totales ┆ precio_petroleo_promedio ┆ var_petroleo ┆ var_ventas │
│ ---  ┆ --- ┆ ---            ┆ ---                      ┆ ---          ┆ ---        │
│ i32  ┆ i8  ┆ f64            ┆ f64                      ┆ f64          ┆ f64        │
╞══════╪═════╪════════════════╪══════════════════════════╪══════════════╪════════════╡
│ 2015 ┆ 1   ┆ 1.4897e7       ┆ 47.609731                ┆ null         ┆ null       │
│ 2015 ┆ 2   ┆ 1.3742e7       ┆ 50.825357                ┆ 0.067541     ┆ -0.077501  │
│ 2015 ┆ 3   ┆ 1.5599e7       ┆ 47.775914                ┆ -0.059998    ┆ 0.135072   │
│ 2015 ┆ 4   ┆ 1.4955e7       ┆ 54.084167                ┆ 0.132038     ┆ -0.041256  │
│ 2015 ┆ 5   ┆ 1.7730e7       ┆ 59.221774                ┆ 0.094993     ┆ 0.185576   │
│ …    ┆ …   ┆ …              ┆ …                        ┆ …            ┆ …          │
│ 2016 ┆ 8   ┆ 2.2452e7     

> **Interpretación:**
> 
> Podemos ver que el lag con mayor correlación es de 1 mes (0.25), aunque es una 
> relación débil. Esto sugiere que la caída del petróleo en 2015-2016 no se refleja 
> de forma clara ni inmediata en las ventas, al menos no con un rezago fijo de meses.


- ¿Qué ciudades mostraron mayor sensibilidad a la caída del petróleo?

In [ ]:
ventas_ciudad_petroleo = (
    ventas_completo
    .filter((pl.col("date").dt.year() >= 2015) & (pl.col("date").dt.year() <= 2016))
    .with_columns([
        pl.col("date").dt.year().alias("anio"),
        pl.col("date").dt.month().alias("mes"),
    ])
    .group_by(["city", "anio", "mes"])
    .agg([
        pl.col("sales").sum().alias("ventas_totales"),
        pl.col("dcoilwtico").mean().alias("precio_petroleo_promedio")
    ])
    .sort(["city", "anio", "mes"])
)

correlacion_por_ciudad = (
    ventas_ciudad_petroleo
    .group_by("city")
    .agg(pl.corr("ventas_totales", "precio_petroleo_promedio").alias("correlacion"))
    .sort("correlacion")
)
print("Ciudades más sensibles (correlación más negativa = ventas suben cuando el petróleo baja):")
print(correlacion_por_ciudad)

Ciudades más sensibles (correlación más negativa = ventas suben cuando el petróleo baja):
shape: (22, 2)
┌───────────────┬─────────────┐
│ city          ┆ correlacion │
│ ---           ┆ ---         │
│ cat           ┆ f64         │
╞═══════════════╪═════════════╡
│ Puyo          ┆ -0.484938   │
│ Santo Domingo ┆ -0.474956   │
│ Cuenca        ┆ -0.455755   │
│ Salinas       ┆ -0.423396   │
│ Guaranda      ┆ -0.414229   │
│ …             ┆ …           │
│ Riobamba      ┆ -0.217178   │
│ Loja          ┆ -0.16496    │
│ Machala       ┆ -0.16242    │
│ Esmeraldas    ┆ -0.088468   │
│ Manta         ┆ -0.0695     │
└───────────────┴─────────────┘


> **Interpretación:**
> 
> Podemos ver que Puyo, Santo Domingo y Cuenca son las ciudades más sensibles a la 
> caída del petróleo (correlación más negativa). En cambio, ciudades como Manta y 
> Esmeraldas casi no muestran relación entre el precio del petróleo y sus ventas.

## Transacciones

- Relación entre número de transacciones y volumen de ventas por tienda.

In [ ]:
ventas_transacciones_tienda = (
    ventas_completo
    .group_by("store_nbr")
    .agg([
        pl.col("sales").sum().alias("ventas_totales"),
        pl.col("transactions").sum().alias("transacciones_totales")
    ])
)

correlacion_ventas_transacciones = ventas_transacciones_tienda.select(
    pl.corr("ventas_totales", "transacciones_totales").alias("correlacion")
)
print(ventas_transacciones_tienda)
print(f"\nCorrelación ventas vs transacciones: {correlacion_ventas_transacciones[0,0]:.4f}")

shape: (54, 3)
┌───────────┬────────────────┬───────────────────────┐
│ store_nbr ┆ ventas_totales ┆ transacciones_totales │
│ ---       ┆ ---            ┆ ---                   │
│ i64       ┆ f64            ┆ i64                   │
╞═══════════╪════════════════╪═══════════════════════╡
│ 10        ┆ 9.6139e6       ┆ 54532269              │
│ 23        ┆ 1.1651e7       ┆ 59009643              │
│ 28        ┆ 1.8383e7       ┆ 64600602              │
│ 32        ┆ 5.9518e6       ┆ 35152458              │
│ 40        ┆ 1.8396e7       ┆ 71979435              │
│ …         ┆ …              ┆ …                     │
│ 43        ┆ 1.6392e7       ┆ 72017748              │
│ 44        ┆ 6.2088e7       ┆ 240012069             │
│ 13        ┆ 1.0523e7       ┆ 51892632              │
│ 50        ┆ 2.8653e7       ┆ 144686652             │
│ 54        ┆ 1.1057e7       ┆ 47892570              │
└───────────┴────────────────┴───────────────────────┘

Correlación ventas vs transacciones: 0.9529


> **Interpretación:**
> 
> Podemos ver una correlación muy fuerte (0.95) entre transacciones y ventas por 
> tienda. Esto indica que, en general, mientras más transacciones tiene una tienda, 
> más ventas totales genera, sin grandes excepciones.

- Identificación de tiendas con ticket promedio alto (pocas transacciones, altas ventas) vs ticket bajo (muchas transacciones, bajas ventas).

In [ ]:
ticket_promedio = (
    ventas_transacciones_tienda
    .with_columns(
        (pl.col("ventas_totales") / pl.col("transacciones_totales")).round(2).alias("ticket_promedio")
    )
    .sort("ticket_promedio", descending=True)
)

print("Top 10 tiendas con TICKET PROMEDIO más alto (pocas transacciones, altas ventas):")
print(ticket_promedio.head(10))

print("\nTop 10 tiendas con TICKET PROMEDIO más bajo (muchas transacciones, bajas ventas):")
print(ticket_promedio.tail(10).sort("ticket_promedio"))

Top 10 tiendas con TICKET PROMEDIO más alto (pocas transacciones, altas ventas):
shape: (10, 4)
┌───────────┬────────────────┬───────────────────────┬─────────────────┐
│ store_nbr ┆ ventas_totales ┆ transacciones_totales ┆ ticket_promedio │
│ ---       ┆ ---            ┆ ---                   ┆ ---             │
│ i64       ┆ f64            ┆ i64                   ┆ f64             │
╞═══════════╪════════════════╪═══════════════════════╪═════════════════╡
│ 51        ┆ 3.2911e7       ┆ 94829262              ┆ 0.35            │
│ 42        ┆ 8.9458e6       ┆ 26487318              ┆ 0.34            │
│ 21        ┆ 9.2555e6       ┆ 27815403              ┆ 0.33            │
│ 52        ┆ 2.6962e6       ┆ 9087969               ┆ 0.3             │
│ 29        ┆ 9.7252e6       ┆ 32387784              ┆ 0.3             │
│ 53        ┆ 1.1216e7       ┆ 38942706              ┆ 0.29            │
│ 3         ┆ 5.0482e7       ┆ 177089550             ┆ 0.29            │
│ 49        ┆ 4.3420e7      

> **Interpretación:**
> 
> Podemos ver que hay tiendas con ticket promedio alto, como la 51 (0.35), donde 
> pocas transacciones generan muchas ventas. En cambio, tiendas como la 34 (0.13) 
> tienen muchas transacciones pero un ticket promedio bajo.

## TAREA 7: EXPORTACIÓN A POSTGRES

In [ ]:
from sqlalchemy import create_engine

recopilacion= {
    "ventas_por_familia": ventas_por_familia,   
    "ventas_por_tienda":ventas_por_tienda,
    "ventas_por_ciudad_provincia":ventas_por_ciudad_provincia, 
    "ventas_mensual":ventas_mensual,
    "ventas_anual":ventas_anual,
    "comparacion_feriados":comparacion_feriados,
    "ventas_ventana_feriado":ventas_ventana_feriado,
    "venta_normal_por_familia":venta_normal_por_familia,
    "venta_feriado_por_familia":venta_feriado_por_familia,
    "sensibilidad_feriados":sensibilidad_feriados,
    "ventas_promocion":ventas_promocion,
    "ventas_petroleo_mensual":ventas_petroleo_mensual,
    "ventas_petroleo_mensual":ventas_petroleo_mensual,
    "periodo_analisis":periodo_analisis,
    "ventas_ciudad_petroleo":ventas_ciudad_petroleo,
    "ventas_transacciones_tienda":ventas_transacciones_tienda,
    "ticket_promedio":ticket_promedio
}

def exportar_postgres(recopilacion):
    engine = create_engine(
        "postgresql+psycopg2://postgres:1234@localhost:5432/proyectoAnalisis"
    )

    for nombre_tabla, df in recopilacion.items():
        df.to_pandas().to_sql(
            nombre_tabla,
            engine,
            if_exists="replace",
            index=False
        )
        print(f"Tabla '{nombre_tabla}' exportada correctamente.")


exportar_postgres(recopilacion)